numba/np/ufunc/tbbpool.cpp

```diff
@@ -217,9 +228,18 @@ static void prepare_fork(void)
     }
     if(tsi)
     {
-        assertion_handler_type orig = tbb::set_assertion_handler(ignore_blocking_terminate_assertion);
-        TSI_TERMINATE(tsi);
-        tbb::set_assertion_handler(orig);
+        if(is_main_thread())
+        {
+            // TBB thread termination must always be called from same thread that called initialize
+            assertion_handler_type orig = tbb::set_assertion_handler(ignore_blocking_terminate_assertion);
+            tsi->blocking_terminate(std::nothrow);
+            tbb::set_assertion_handler(orig);
+            need_reinit_after_fork = true;
+        }
+        else if (_DEBUG)
+        {
+            puts("NUMBA: Attempt to fork from non-main thread, TBB may be in incorrect state in child processes");
```

Comment:
Perhaps, make it a hard error (abrt?) and inform user to start TBB pool in the main thread.

Reply:
It will abort on Stuart repro from #5973 then, because `multiprocessing` create separate thread internally to launch new processes

Reply:
The issue in that example that `multiprocessing ` uses `popen` to create new processes and this is completely fine (as popen creates completely new process), but it is still invoking fork handlers and we can't reliably detect this from `atfork`.

Also,
https://sourceware.org/bugzilla/show_bug.cgi?id=17490
it seems that `popen` behavior was changed in recent glibc versions but we still can have `fork` + `exec` pair, which is completely fine by POSIX standard, completely safe (as `exec` replaces child process completely), and no way to reliably detect this.

Reply:
Given there's no way to easy way to detect `fork` vs `fork + exec` I think the debug print should be on by default and print to `stderr`. That no one has hit this so far suggests it's probably a rare use case and my preference would be to report it to users and then see what sort of feedback (if any) comes in. I've made a patch that implements this change and also adds a test for it here: https://github.com/stuartarchibald/numba/commit/395f2ae6865df5b9a2f5c6001f700e64b70844d7 if you think it's ok then perhaps add it on to this PR? Thanks.

Reply:
Cherry-picked, thanks